src/producer/mod.rs

```diff
@@ -195,12 +195,58 @@ pub trait ProducerContext: ClientContext {
     /// method once the message has been delivered, or failed to.
     type DeliveryOpaque: IntoOpaque;
 
+    /// Partitioner is a user-defined structure that will called
+    /// when deciding to which partition send a message to.
+    ///
+    /// Associated type defaults are unstable (https://github.com/rust-lang/rust/issues/29661) so
+    /// for now it isn't possible to set it to NoCustomPartitioner.
+    type CustomPartitioner: Partitioner;
```

Comment:
maybe we should use generic type parameter instead with default value as in std::ops::Add ?